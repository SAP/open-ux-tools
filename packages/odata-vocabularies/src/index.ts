export * from './types/vocabulary-service';
export { VocabularyNamespace, VocabularyAlias } from './resources/index';
export { Namespace, COMPLEX_TYPE_KIND, ENUM_TYPE_KIND, TERM_KIND, TYPE_DEFINITION_KIND } from './types/base-types';
export { VocabularyService } from './vocabulary-service';
