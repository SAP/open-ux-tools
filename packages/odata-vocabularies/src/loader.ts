import type {
    CSDLAnnotations,
    UnboxContent,
    TypeDefinition as CSDLTypeDefinition,
    SchemaElement,
    EnumTypeBase,
    ComplexType as CSDLComplexType,
    Term as CSDLTerm,
    CSDL
} from '@sap-ux/vocabularies/CSDL';

import type { VocabularyAlias, VocabularyNamespace } from './resources';
import VOCABULARIES, { ALIAS_TO_NAMESPACE, NAMESPACE_TO_ALIAS } from './resources';
import type {
    EdmNameType,
    Constraints,
    BaseVocabularyObject,
    ComplexType,
    EnumType,
    EnumValue,
    Facets,
    PrimitiveType,
    Term,
    TypeDefinition,
    VocabularyObject,
    TargetKindValue,
    ComplexTypeProperty,
    Vocabulary,
    AllowedValues,
    VocabulariesInformation
} from './types';
import {
    ENUM_VALUE_KIND,
    PROPERTY_KIND,
    CDS_VOCABULARY_ALIAS,
    CDS_VOCABULARY_NAMESPACE,
    TYPE_DEFINITION_KIND,
    ENUM_TYPE_KIND,
    COMPLEX_TYPE_KIND,
    TERM_KIND
} from './types';

const PROPERTY_PATTERN = /^(\w){1,128}$/;

const SUPPORTED_VOCABULARY_NAMESPACES: Set<VocabularyNamespace> = new Set([
    'Org.OData.Aggregation.V1',
    'Org.OData.Authorization.V1',
    'Org.OData.Capabilities.V1',
    'Org.OData.Core.V1',
    'Org.OData.Measures.V1',
    'Org.OData.Repeatability.V1',
    'Org.OData.Temporal.V1',
    'Org.OData.Validation.V1',
    'Org.OData.JSON.V1',
    'com.sap.vocabularies.Analytics.v1',
    'com.sap.vocabularies.CDS.v1',
    'com.sap.vocabularies.CodeList.v1',
    'com.sap.vocabularies.Common.v1',
    'com.sap.vocabularies.Communication.v1',
    'com.sap.vocabularies.DataIntegration.v1',
    'com.sap.vocabularies.DirectEdit.v1',
    'com.sap.vocabularies.Graph.v1',
    'com.sap.vocabularies.ODM.v1',
    'com.sap.vocabularies.PDF.v1',
    'com.sap.vocabularies.PersonalData.v1',
    'com.sap.vocabularies.Hierarchy.v1',
    'com.sap.vocabularies.Session.v1',
    'com.sap.vocabularies.UI.v1',
    'com.sap.vocabularies.HTML5.v1'
]);

let vocabulariesInformationStatic: VocabulariesInformation | undefined;
let vocabulariesInformationStaticCds: VocabulariesInformation | undefined;

/**
 *
 * @param base Vocabulary object
 * @param raw CSDL object
 * @param prefix Annotation prefix
 */
function fillBaseProperties<T extends UnboxContent<CSDLAnnotations>>(
    base: BaseVocabularyObject,
    raw: T,
    prefix: string = ''
): void {
    const description = raw[`${prefix}@Org.OData.Core.V1.Description`];
    if (description !== undefined) {
        base.description = description;
    }
    const longDescription = raw[`${prefix}@Org.OData.Core.V1.LongDescription`];
    if (longDescription !== undefined) {
        base.longDescription = longDescription;
    }
    const experimental = raw[`${prefix}@com.sap.vocabularies.Common.v1.Experimental`];
    if (experimental !== undefined) {
        base.experimental = experimental;
    }

    // deprecated status
    const revisions = raw[`${prefix}@Org.OData.Core.V1.Revisions`] ?? [];
    const deprecatedRevisions = revisions.find(
        (revision: { Kind: string; Description: string }) => revision.Kind === 'Deprecated'
    );
    if (deprecatedRevisions) {
        base.deprecated = true;
        base.deprecatedDescription = deprecatedRevisions.Description;
    }
}

/**
 *
 * @param primitiveType Vocabulary Object
 * @param raw CSDL Object
 */
function fillPrimitiveTypeProperties(primitiveType: PrimitiveType, raw: SchemaElement): void {
    fillBaseProperties(primitiveType, raw);
    if (raw.$UnderlyingType !== undefined) {
        primitiveType.underlyingType = raw.$UnderlyingType;
    }
}

/**
 *
 * @param raw CSDL Object
 * @returns Facets
 */
function getFacets(raw: SchemaElement): Facets | undefined {
    const facets: Facets = {};
    if (raw.$Nullable !== undefined) {
        facets.isNullable = !!raw.$Nullable;
    }
    if (raw.$Precision !== undefined && Number.isInteger(raw.$Precision)) {
        facets.precision = raw.$Precision;
    }
    return Object.keys(facets).length ? facets : undefined;
}

/**
 *
 * @param raw CSDL Object
 * @returns Allowed values
 */
function parseAllowedValues(raw: UnboxContent<CSDLAnnotations>): AllowedValues[] {
    const rawAllowedValue = raw['@Org.OData.Validation.V1.AllowedValues'];
    if (rawAllowedValue?.length) {
        return rawAllowedValue.map((rawValue: UnboxContent<CSDLAnnotations> & { Value: unknown }) => ({
            value: rawValue.Value,
            description: rawValue['@Org.OData.Core.V1.Description'],
            longDescription: rawValue['@Org.OData.Core.V1.LongDescription'] ?? ''
        }));
    }
    return [];
}

/**
 *
 * @param key string
 * @returns boolean
 */
const isValidKey = (key: string) => {
    const maxLength = 512;
    if (key.startsWith('$') && key.length <= maxLength) {
        return true;
    }

    if (key.indexOf('@') >= 0 && key.length <= maxLength) {
        return true;
    }

    return false;
};

/**
 *
 * @param raw CSDL Object
 * @returns Constraints
 */
function getConstraints(raw: UnboxContent<CSDLAnnotations>): Constraints {
    const constraints: Constraints = {};
    const allowedValues: AllowedValues[] = parseAllowedValues(raw);
    if (allowedValues?.length) {
        constraints.allowedValues = allowedValues;
    }

    const allowedTerms = raw['@Org.OData.Validation.V1.AllowedTerms'];
    if (allowedTerms?.length) {
        constraints.allowedTerms = allowedTerms.map(getFullyQualifiedAllowedTermName);
    }

    const isLanguageDependent = raw['@Org.OData.Core.V1.IsLanguageDependent'];
    if (isLanguageDependent !== undefined) {
        constraints.isLanguageDependent = !!isLanguageDependent;
    }

    const requiresType = raw['@Org.OData.Core.V1.RequiresType'];
    if (requiresType) {
        constraints.requiresType = requiresType;
    }
    const openPropertyTypeConstraints = raw['@Org.OData.Validation.V1.OpenPropertyTypeConstraint'];
    if (openPropertyTypeConstraints) {
        constraints.openPropertyTypeConstraints = openPropertyTypeConstraints;
    }
    return constraints;
}

/**
 *
 * @param name Name of the type definition
 * @param raw CSDL Object
 * @returns Vocabulary Object
 */
function parseTypeDefinition(name: string, raw: CSDLTypeDefinition): TypeDefinition {
    const typeDef: TypeDefinition = { kind: TYPE_DEFINITION_KIND, name: name };
    fillPrimitiveTypeProperties(typeDef, raw);
    const facets = getFacets(raw);
    if (facets) {
        typeDef.facets = facets;
    }
    // constraints
    const constraints = getConstraints(raw);
    if (Object.keys(constraints).length) {
        typeDef.constraints = constraints;
    }
    return typeDef;
}

/**
 *
 * @param name Name of the enum
 * @param raw CSDL Object
 * @returns Vocabulary Object
 */
export function parseEnumTypeDefinition(name: string, raw: EnumTypeBase): EnumType {
    const enumType: EnumType = {
        kind: ENUM_TYPE_KIND,
        name: name,
        isFlags: false,
        values: [],
        deprecated: false
    };
    try {
        fillPrimitiveTypeProperties(enumType, raw);
        enumType.isFlags = raw.$IsFlags === true;

        // scan for enum values
        enumType.values = Object.keys(raw)
            .filter((key) => !isValidKey(key))
            .map((key) => {
                const value = parseInt(raw[key], 10);
                const enumValue: EnumValue = {
                    kind: ENUM_VALUE_KIND,
                    name: key,
                    value
                };
                fillBaseProperties(enumValue, raw, key);
                return enumValue;
            });
    } catch (e) {
        return enumType;
    }
    return enumType;
}

/**
 *
 * @param name Name of the complex type
 * @param raw CSDL Object
 * @returns Vocabulary Object
 */
function parseComplexType(name: string, raw: CSDLComplexType): ComplexType {
    const complexType: ComplexType = { kind: COMPLEX_TYPE_KIND, name: name, properties: new Map() };
    fillPrimitiveTypeProperties(complexType, raw);
    if (raw.$BaseType !== undefined) {
        complexType.baseType = raw.$BaseType;
    }
    if (raw.$Abstract !== undefined) {
        complexType.isAbstract = !!raw.$Abstract;
    }
    if (raw.$OpenType !== undefined) {
        complexType.isOpenType = !!raw.$OpenType;
    }

    // collect properties
    Object.keys(raw)
        .filter((key) => !isValidKey(key))
        .forEach((key) => {
            const propRaw = raw[key];
            const property: ComplexTypeProperty = {
                kind: PROPERTY_KIND,
                name: key,
                type: propRaw.$Type ?? 'Edm.String',
                isCollection: !!propRaw.$Collection
            };
            fillBaseProperties(property, propRaw);
            if (propRaw.$DefaultValue !== undefined) {
                property.defaultValue = propRaw.$DefaultValue;
            }
            // facets
            const facets = getFacets(propRaw);
            if (facets) {
                property.facets = facets;
            }
            // constraints
            const constraints = getConstraints(propRaw);
            const derivedTypeConstraints = propRaw['@Org.OData.Validation.V1.DerivedTypeConstraint'];
            if (derivedTypeConstraints?.length) {
                constraints.derivedTypeConstraints = derivedTypeConstraints;
            }
            if (Object.keys(constraints).length) {
                property.constraints = constraints;
            }
            complexType.properties.set(property.name, property);
        });
    return complexType;
}

/**
 *
 * @param name Name of the term
 * @param raw CSDL Object
 * @returns Vocabulary Object
 */
function parseTerm(name: string, raw: CSDLTerm): Term {
    const term: Term = {
        kind: TERM_KIND,
        name: name,
        type: raw.$Type ?? 'Edm.String',
        isCollection: !!raw.$Collection
    };
    fillBaseProperties(term, raw);
    if (raw.$DefaultValue !== undefined) {
        term.defaultValue = raw.$DefaultValue;
    }
    if (raw.$AppliesTo !== undefined) {
        term.appliesTo = raw.$AppliesTo;
    }
    if (raw.$BaseTerm !== undefined) {
        term.baseTerm = raw.$BaseTerm;
    }
    if (raw.$CdsName !== undefined) {
        term.cdsName = raw.$CdsName;
    }

    // facets
    const facets = getFacets(raw);
    if (facets) {
        term.facets = facets;
    }
    // constraints
    const constraints = getConstraints(raw);
    if (Object.keys(constraints).length) {
        term.constraints = constraints;
    }
    return term;
}

/**
 *
 * @param identifier Name
 * @param element CSDL Object
 * @returns Vocabulary Object
 */
function parseSchemaElements(identifier: string, element: SchemaElement): VocabularyObject | undefined {
    if (element.$Kind !== undefined) {
        switch (element.$Kind) {
            case 'EnumType':
                return parseEnumTypeDefinition(identifier, element);

            case 'TypeDefinition':
                return parseTypeDefinition(identifier, element);

            case 'ComplexType':
                return parseComplexType(identifier, element);

            case 'Term':
                return parseTerm(identifier, element);

            default:
        }
    }
    return undefined;
}

/**
 *
 * @param includeCds Flag indicating if CDS vocabularies should be loaded
 * @returns Vocabularies
 */
export const loadVocabulariesInformation = (includeCds?: boolean): VocabulariesInformation => {
    // try to use cache
    if (includeCds && vocabulariesInformationStaticCds) {
        return vocabulariesInformationStaticCds;
    } else if (!includeCds && vocabulariesInformationStatic) {
        return vocabulariesInformationStatic;
    }

    const dictionary: Map<EdmNameType.FullyQualifiedName, VocabularyObject> = new Map();
    const byTarget: Map<TargetKindValue | '', Set<EdmNameType.FullyQualifiedName>> = new Map();
    const supportedVocabularies: Map<VocabularyNamespace, Vocabulary> = new Map();
    const namespaceByDefaultAlias: Map<EdmNameType.SimpleIdentifier, VocabularyNamespace> = new Map();
    const derivedTypesPerType: Map<
        EdmNameType.FullyQualifiedName,
        Map<EdmNameType.FullyQualifiedName, boolean>
    > = new Map();
    const upperCaseNameMap: Map<string, string | Map<string, string>> = new Map();

    const addToUpperCaseNameMap = (name: string, propertyName?: string): void => {
        const upperCaseName = name.toUpperCase();
        const currentEntry = upperCaseNameMap.get(upperCaseName);
        if (!propertyName) {
            if (!currentEntry || typeof currentEntry === 'string') {
                upperCaseNameMap.set(upperCaseName, name);
            } else if (typeof currentEntry === 'object') {
                upperCaseNameMap.set(upperCaseName, Object.assign(currentEntry, { $Self: name }));
            }
        } else {
            const upperCasePropName = propertyName.toUpperCase();
            if (!currentEntry) {
                upperCaseNameMap.set(upperCaseName, new Map([['$Self', name]]));
            } else if (typeof currentEntry === 'string') {
                upperCaseNameMap.set(upperCaseName, new Map([['$Self', currentEntry]]));
            }
            (upperCaseNameMap.get(upperCaseName) as Map<string, string>).set(upperCasePropName, propertyName);
        }
    };

    const getVocabularyUri = (vocabulary: any): string => {
        const links: { rel?: string; href?: string }[] = vocabulary?.['@Org.OData.Core.V1.Links'] ?? [];
        const linkMap: Map<string, string> = new Map();
        for (const link of links) {
            linkMap.set(link.rel ?? '', link.href ?? '');
        }
        // get url pointing to latest version xml - if not available use alternative
        const latestVersion = linkMap.get('latest-version');
        const alternate = linkMap.get('alternate');
        if (latestVersion?.endsWith('.xml')) {
            return latestVersion;
        } else if (alternate?.endsWith('.xml')) {
            return alternate;
        } else if (latestVersion) {
            return latestVersion;
        } else if (alternate) {
            return alternate;
        }
        return '';
    };

    NAMESPACE_TO_ALIAS.forEach((alias, namespace) => {
        if (!includeCds && alias === CDS_VOCABULARY_ALIAS) {
            return;
        }
        addToUpperCaseNameMap(alias);
        addToUpperCaseNameMap(namespace);
        namespaceByDefaultAlias.set(alias, namespace);
    });

    for (const namespace of SUPPORTED_VOCABULARY_NAMESPACES) {
        if (!includeCds && namespace === CDS_VOCABULARY_NAMESPACE) {
            continue;
        }
        const alias = NAMESPACE_TO_ALIAS.get(namespace);
        if (alias) {
            const document: CSDL = VOCABULARIES[alias];
            if (document) {
                const vocabulary = document[namespace];
                supportedVocabularies.set(namespace, {
                    namespace: namespace,
                    defaultAlias: alias,
                    defaultUri: getVocabularyUri(vocabulary)
                });
                const properties = Object.keys(vocabulary);
                for (const identifier of properties.filter((property) => PROPERTY_PATTERN.test(property))) {
                    const fqName = namespace + '.' + identifier;
                    const element = vocabulary[identifier];
                    const vocabularyObject = parseSchemaElements(fqName, element);
                    if (vocabularyObject) {
                        dictionary.set(fqName, vocabularyObject);
                        addToUpperCaseNameMap(fqName);
                        if (vocabularyObject.kind === TERM_KIND) {
                            for (const targetKind of vocabularyObject.appliesTo ?? ['']) {
                                byTarget.set(targetKind, byTarget.get(targetKind) ?? new Set());
                                byTarget.get(targetKind)?.add(fqName);
                            }
                        } else if (vocabularyObject.kind === COMPLEX_TYPE_KIND) {
                            vocabularyObject.properties.forEach((property) => {
                                addToUpperCaseNameMap(fqName, property.name);
                            });
                            const baseType = vocabularyObject.baseType;
                            if (baseType) {
                                const baseName = baseType.split('.').pop() ?? '';
                                const baseElement = vocabulary[baseName];
                                if (baseElement) {
                                    const vocabularyBaseObject = parseSchemaElements(baseType, baseElement);
                                    // add property of baseType to fqName
                                    if (vocabularyBaseObject?.kind === COMPLEX_TYPE_KIND) {
                                        vocabularyBaseObject.properties.forEach((property) => {
                                            addToUpperCaseNameMap(fqName, property.name);
                                        });
                                    }
                                }
                                derivedTypesPerType.set(baseType, derivedTypesPerType.get(baseType) ?? new Map());
                                derivedTypesPerType.get(baseType)?.set(fqName, !!vocabularyObject.isAbstract);
                            }
                        }
                        if (vocabularyObject.kind === ENUM_TYPE_KIND) {
                            vocabularyObject.values.forEach((value) => {
                                addToUpperCaseNameMap(fqName, value.name);
                            });
                        }
                    }
                }
            }
        }
    }
    const vocabulariesInformation: VocabulariesInformation = {
        dictionary,
        byTarget,
        derivedTypesPerType,
        supportedVocabularies,
        namespaceByDefaultAlias,
        upperCaseNameMap
    };
    // fill cache
    if (includeCds) {
        vocabulariesInformationStaticCds = vocabulariesInformation;
    } else {
        vocabulariesInformationStatic = vocabulariesInformation;
    }
    return vocabulariesInformation;
};

/**
 *
 * @param allowedTerm Term name
 * @returns fully qualified term name
 */
function getFullyQualifiedAllowedTermName(allowedTerm: string): string {
    const segments = allowedTerm.split('.');
    if (segments.length !== 2) {
        return allowedTerm;
    }
    const namespace = ALIAS_TO_NAMESPACE.get(segments[0] as VocabularyAlias);
    if (!namespace) {
        return allowedTerm;
    }
    segments[0] = namespace;
    return segments.join('.');
}
