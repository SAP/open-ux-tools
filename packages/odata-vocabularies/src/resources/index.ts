import type { CSDL } from '@sap-ux/vocabularies/CSDL';

import Aggregation from './Org.OData.Aggregation.V1';
import Auth from './Org.OData.Authorization.V1';
import Capabilities from './Org.OData.Capabilities.V1';
import Core from './Org.OData.Core.V1';
import Measures from './Org.OData.Measures.V1';
import Repeatability from './Org.OData.Repeatability.V1';
import Temporal from './Org.OData.Temporal.V1';
import Validation from './Org.OData.Validation.V1';
import ODataJSON from './Org.OData.JSON.V1';
import Analytics from './com.sap.vocabularies.Analytics.v1';
import CDS from './com.sap.vocabularies.CDS.v1';
import CodeList from './com.sap.vocabularies.CodeList.v1';
import Common from './com.sap.vocabularies.Common.v1';
import Communication from './com.sap.vocabularies.Communication.v1';
import DataIntegration from './com.sap.vocabularies.DataIntegration.v1';
import DirectEdit from './com.sap.vocabularies.DirectEdit.v1';
import Graph from './com.sap.vocabularies.Graph.v1';
import Hierarchy from './com.sap.vocabularies.Hierarchy.v1';
import ODM from './com.sap.vocabularies.ODM.v1';
import PDF from './com.sap.vocabularies.PDF.v1';
import PersonalData from './com.sap.vocabularies.PersonalData.v1';
import Session from './com.sap.vocabularies.Session.v1';
import UI from './com.sap.vocabularies.UI.v1';
import HTML5 from './com.sap.vocabularies.HTML5.v1';

export type SapVocabularyNamespace =
    | 'com.sap.vocabularies.Analytics.v1'
    | 'com.sap.vocabularies.CDS.v1'
    | 'com.sap.vocabularies.CodeList.v1'
    | 'com.sap.vocabularies.Common.v1'
    | 'com.sap.vocabularies.Communication.v1'
    | 'com.sap.vocabularies.DataIntegration.v1'
    | 'com.sap.vocabularies.DirectEdit.v1'
    | 'com.sap.vocabularies.Graph.v1'
    | 'com.sap.vocabularies.Hierarchy.v1'
    | 'com.sap.vocabularies.ODM.v1'
    | 'com.sap.vocabularies.PDF.v1'
    | 'com.sap.vocabularies.PersonalData.v1'
    | 'com.sap.vocabularies.Session.v1'
    | 'com.sap.vocabularies.UI.v1'
    | 'com.sap.vocabularies.HTML5.v1';

export type SapVocabularyAlias =
    | 'Analytics'
    | 'CDS'
    | 'CodeList'
    | 'Common'
    | 'Communication'
    | 'DataIntegration'
    | 'DirectEdit'
    | 'Graph'
    | 'Hierarchy'
    | 'ODM'
    | 'PDF'
    | 'PersonalData'
    | 'Session'
    | 'UI'
    | 'HTML5';

export type OasisVocabularyNamespace =
    | 'Org.OData.Aggregation.V1'
    | 'Org.OData.Authorization.V1'
    | 'Org.OData.Capabilities.V1'
    | 'Org.OData.Core.V1'
    | 'Org.OData.Measures.V1'
    | 'Org.OData.Repeatability.V1'
    | 'Org.OData.Temporal.V1'
    | 'Org.OData.Validation.V1'
    | 'Org.OData.JSON.V1';

export type OasisVocabularyAlias =
    | 'Aggregation'
    | 'Auth'
    | 'Capabilities'
    | 'Core'
    | 'Measures'
    | 'Repeatability'
    | 'Temporal'
    | 'Validation'
    | 'ODataJSON';

export type VocabularyNamespace = SapVocabularyNamespace | OasisVocabularyNamespace;
export type VocabularyAlias = SapVocabularyAlias | OasisVocabularyAlias;

export const NAMESPACE_TO_ALIAS: Map<VocabularyNamespace, VocabularyAlias> = new Map([
    ['Org.OData.Aggregation.V1', 'Aggregation'],
    ['Org.OData.Authorization.V1', 'Auth'],
    ['Org.OData.Capabilities.V1', 'Capabilities'],
    ['Org.OData.Core.V1', 'Core'],
    ['Org.OData.Measures.V1', 'Measures'],
    ['Org.OData.Repeatability.V1', 'Repeatability'],
    ['Org.OData.Temporal.V1', 'Temporal'],
    ['Org.OData.Validation.V1', 'Validation'],
    ['Org.OData.JSON.V1', 'ODataJSON'],
    ['com.sap.vocabularies.Analytics.v1', 'Analytics'],
    ['com.sap.vocabularies.CDS.v1', 'CDS'],
    ['com.sap.vocabularies.CodeList.v1', 'CodeList'],
    ['com.sap.vocabularies.Common.v1', 'Common'],
    ['com.sap.vocabularies.Communication.v1', 'Communication'],
    ['com.sap.vocabularies.DataIntegration.v1', 'DataIntegration'],
    ['com.sap.vocabularies.DirectEdit.v1', 'DirectEdit'],
    ['com.sap.vocabularies.Graph.v1', 'Graph'],
    ['com.sap.vocabularies.Hierarchy.v1', 'Hierarchy'],
    ['com.sap.vocabularies.ODM.v1', 'ODM'],
    ['com.sap.vocabularies.PDF.v1', 'PDF'],
    ['com.sap.vocabularies.PersonalData.v1', 'PersonalData'],
    ['com.sap.vocabularies.Session.v1', 'Session'],
    ['com.sap.vocabularies.UI.v1', 'UI'],
    ['com.sap.vocabularies.HTML5.v1', 'HTML5']
]);

export const ALIAS_TO_NAMESPACE: Map<VocabularyAlias, VocabularyNamespace> = new Map([
    ['Aggregation', 'Org.OData.Aggregation.V1'],
    ['Auth', 'Org.OData.Authorization.V1'],
    ['Capabilities', 'Org.OData.Capabilities.V1'],
    ['Core', 'Org.OData.Core.V1'],
    ['Measures', 'Org.OData.Measures.V1'],
    ['Repeatability', 'Org.OData.Repeatability.V1'],
    ['Temporal', 'Org.OData.Temporal.V1'],
    ['Validation', 'Org.OData.Validation.V1'],
    ['ODataJSON', 'Org.OData.JSON.V1'],
    ['Analytics', 'com.sap.vocabularies.Analytics.v1'],
    ['CDS', 'com.sap.vocabularies.CDS.v1'],
    ['CodeList', 'com.sap.vocabularies.CodeList.v1'],
    ['Common', 'com.sap.vocabularies.Common.v1'],
    ['Communication', 'com.sap.vocabularies.Communication.v1'],
    ['DataIntegration', 'com.sap.vocabularies.DataIntegration.v1'],
    ['DirectEdit', 'com.sap.vocabularies.DirectEdit.v1'],
    ['Graph', 'com.sap.vocabularies.Graph.v1'],
    ['Hierarchy', 'com.sap.vocabularies.Hierarchy.v1'],
    ['ODM', 'com.sap.vocabularies.ODM.v1'],
    ['PDF', 'com.sap.vocabularies.PDF.v1'],
    ['PersonalData', 'com.sap.vocabularies.PersonalData.v1'],
    ['Session', 'com.sap.vocabularies.Session.v1'],
    ['UI', 'com.sap.vocabularies.UI.v1'],
    ['HTML5', 'com.sap.vocabularies.HTML5.v1']
]);

const vocabularies: Record<string, CSDL> = {
    Aggregation,
    Auth,
    Capabilities,
    Core,
    Measures,
    Repeatability,
    Temporal,
    Validation,
    ODataJSON,
    Analytics,
    CDS,
    CodeList,
    Common,
    Communication,
    DataIntegration,
    DirectEdit,
    Graph,
    Hierarchy,
    ODM,
    PDF,
    PersonalData,
    Session,
    UI,
    HTML5
};

export default vocabularies;
