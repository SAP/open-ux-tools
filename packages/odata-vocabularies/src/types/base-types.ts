/**
 * common OData types used in vocabulary and metadata API
 */

/**
 * String types used for documentation purposes
 * See symbolic values listed in http://docs.oasis-open.org/odata/odata-csdl-xml/v4.01/odata-csdl-xml-v4.01.html#sec_Applicability
 */
export type TargetKind = string;
/**
 * starts with a letter or underscore, followed by at most 127 letters, underscores or digits
 */
export type SimpleIdentifier = string;
export type Name = SimpleIdentifier;
export type Alias = SimpleIdentifier;
/**
 * dot separated sequence of SimpleIdentifier
 */
export type Namespace = string;
export type NameQualifier = SimpleIdentifier | string; //  <Namespace|Alias>
export type AliasQualifiedName = string; // <Alias>.<Name>
export type FullyQualifiedName = string; // <Namespace>.<Name>
export type QualifiedName = AliasQualifiedName | FullyQualifiedName; // <Namespace|Alias>.<Name>
export type FullyQualifiedTypeName = string; // <Namespace>.<Name> || Collection(<Namespace>.<Name>)

export type TargetPath = string; // target path value for externally targeting annotations

/**
 * Target Kinds
 *
 * used for finding allowed vocabulary terms (kinds are restricted via AppliesTo attribute) to service meta data objects
 */

// kinds which have fully qualified name
export const TYPE_DEFINITION_KIND = 'TypeDefinition';
export const ENUM_TYPE_KIND = 'EnumType';
export const COMPLEX_TYPE_KIND = 'ComplexType';
export const TERM_KIND = 'Term';

export const ENTITY_TYPE_KIND = 'EntityType';

export const ACTION_KIND = 'Action';
export const FUNCTION_KIND = 'Function';

export const ASSOCIATION_KIND = 'Association'; // V2 only

export const ENTITY_CONTAINER_KIND = 'EntityContainer';

// kinds which have simple identifier only
// (embedded in structural types)
export const PROPERTY_KIND = 'Property';
export const NAVIGATION_PROPERTY_KIND = 'NavigationProperty';

// (embedded in entity container)
export const ENTITY_SET_KIND = 'EntitySet';
export const SINGLETON_KIND = 'Singleton';
export const ACTION_IMPORT_KIND = 'ActionImport';
export const FUNCTION_IMPORT_KIND = 'FunctionImport';
export const ASSOCIATION_SET_KIND = 'AssociationSet'; // V2 only

// miscellaneous
export const COLLECTION_KIND = 'Collection';
