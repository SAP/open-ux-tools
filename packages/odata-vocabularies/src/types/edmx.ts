/* eslint-disable @typescript-eslint/no-namespace */
import type { Edm } from './edm';

export namespace Edmx {
    export namespace Json {
        export interface Document {
            $Version: string;
            $Refrence?: any;
            [namespace: string]: Edm.Schema | string | any;
        }
    }
}
