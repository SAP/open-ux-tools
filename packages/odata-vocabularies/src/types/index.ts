export * from './base-types';
export * from './names';
export * from './vocabulary-service';
