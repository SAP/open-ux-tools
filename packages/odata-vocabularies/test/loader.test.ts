import { loadVocabulariesInformation } from './../src/loader';
import type { Term } from '../src/types/vocabulary-service';
import { ENTITY_TYPE_KIND } from '../src/types/base-types';

declare const expect: jest.Expect;

const vocabularyInformation = loadVocabulariesInformation();
const namespace = 'com.sap.vocabularies.UI.v1';
const term = 'com.sap.vocabularies.UI.v1.LineItem';
const target = ENTITY_TYPE_KIND;

describe('lib vocabulary loader', () => {
    it('correct term definition in dictionary', () => {
        // Expect
        expect(vocabularyInformation.dictionary.get(term)).toMatchInlineSnapshot(`
            Object {
              "appliesTo": Array [
                "EntityType",
              ],
              "description": "Collection of data fields for representation in a table or list",
              "isCollection": true,
              "kind": "Term",
              "name": "com.sap.vocabularies.UI.v1.LineItem",
              "type": "com.sap.vocabularies.UI.v1.DataFieldAbstract",
            }
        `);
    });
    it('term listed as available for kind (based on AppliesTo) ', () => {
        // Act
        const specificTargetKind = vocabularyInformation.byTarget.get(target);
        // Expect
        expect(specificTargetKind?.has(term)).toBeTruthy();
    });
    it('namespace listed with alias ', () => {
        // Expect
        expect(vocabularyInformation.supportedVocabularies.get(namespace)).toMatchInlineSnapshot(`
            Object {
              "defaultAlias": "UI",
              "defaultUri": "https://sap.github.io/odata-vocabularies/vocabularies/UI.xml",
              "namespace": "com.sap.vocabularies.UI.v1",
            }
        `);
    });
    it('check derived types for terms type ', () => {
        // Arrange
        const termObject = vocabularyInformation.dictionary.get(term) as Term;
        // Expect
        expect(vocabularyInformation.derivedTypesPerType.get(termObject.type)).toMatchInlineSnapshot(`
            Map {
              "com.sap.vocabularies.UI.v1.DataFieldForAnnotation" => false,
              "com.sap.vocabularies.UI.v1.DataFieldForActionAbstract" => true,
              "com.sap.vocabularies.UI.v1.DataFieldForActionGroup" => false,
              "com.sap.vocabularies.UI.v1.DataField" => false,
            }
        `);
    });
    it('check upper case name is present', () => {
        expect(vocabularyInformation.upperCaseNameMap.get(term.toUpperCase())).toBe(term);
    });

    it('check upper case name considers baseType', () => {
        const term = 'com.sap.vocabularies.UI.v1.ReferenceFacet';
        const result = vocabularyInformation.upperCaseNameMap.get(term.toUpperCase()) as Map<string, string>;
        const id = result.get('ID');
        const label = result.get('LABEL');
        expect(id).toStrictEqual('ID');
        expect(label).toStrictEqual('Label');
    });
});
