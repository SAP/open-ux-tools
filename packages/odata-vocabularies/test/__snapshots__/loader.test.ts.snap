// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib vocabulary loader JSON vocabulary namespace listed with correct alias 1`] = `
Object {
  "defaultAlias": "JSON",
  "defaultUri": "https://oasis-tcs.github.io/odata-vocabularies/vocabularies/Org.OData.JSON.V1.xml",
  "namespace": "Org.OData.JSON.V1",
}
`;
